### PR TITLE
ASoC: sof: bxt_pcm512x adds device_link

### PR DESCRIPTION
**Pull request by libinyang**
In resume from S3, HDAC HDMI codec driver dapm event callback may be
operated before HDMI codec driver turns on the display audio power
domain because of the contest between display driver and hdmi codec driver.

This patch adds the device_link between machine device (consumer) and
hdmi codec device (supplier) to make sure the sequence is always correct.

Signed-off-by: Libin Yang <libin.yang@intel.com>

### CODE REVIEW

**Comment by libinyang** [comment]
As per discussion with Takashi, he likes more the device_link to resolve the hdmi power sequence issue.
I send the patch just for bxt_pcm512X now. And if it passes the review, I will send patches for more machines. Please help review.
I proposed to resolve this issue in consumer driver. No reply from Takashi so far.

File: sound/soc/intel/boards/bxt_pcm512x.c [hunk 1]
```diff
@@ -53,6 +54,15 @@ static int broxton_hdmi_init(struct snd_soc_pcm_runtime *rtd)
 
 	list_add_tail(&pcm->head, &ctx->hdmi_pcm_list);
 
+	/*
+	 * Setup the device_link to make sure that hdmi codec power is
+	 * turned on before the consumer starts to operate on the
+	 * codec register.
+	 */
+	if (!ctx->link)
+		ctx->link = device_link_add(rtd->card->dev, dai->dev,
```
Comment (ranj063) [inline sound/soc/intel/boards/bxt_pcm512x.c:63]: @libinyang looks like we need to check for null after device_link_add as it could return null?

**Comment by ranj063** [comment]
@libinyang just for my education: the race is between the hdmi codec and display driver. Why are we adding the device link between the machine driver and the hdmi device?

File: sound/soc/intel/boards/bxt_pcm512x.c (hunk 1 above)
Comment (libinyang) [reply sound/soc/intel/boards/bxt_pcm512x.c:63]: If it returns NULL, it means the link is not added. But I don't think we should stop the init. Or we add a dev_warn?

**Comment by ranj063** [comment]
Also wondering that the one issue we've discussed before is that the hdmi device and the sof-pci-audio device do not have a parent-child relationship like the other devices like the dmic-codec, machine drv and the component drv. Do you think it makes sense to add a device link between the sof device and the hdmi device?

**Comment by libinyang** [comment]
@ranj063 Takashi educated me :) And I think it's pretty good for such situation. If we add the device link, we can make sure that the power on sequence is correct as they are consumer-supplier. The device link core will make sure the supplier power is always be turned on firstly.

> @libinyang just for my education: the race is between the hdmi codec and display driver. Why are we adding the device link between the machine driver and the hdmi device?

**Comment by ranj063** [comment]
> Also wondering that the one issue we've discussed before is that the hdmi device and the sof-pci-audio device do not have a parent-child relationship like the other devices like the dmic-codec, machine drv and the component drv. Do you think it makes sense to add a device link between the sof device and the hdmi device?

Thanks for clarifying @libinyang. I think I understand the PR better now.

File: sound/soc/intel/boards/bxt_pcm512x.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/intel/boards/bxt_pcm512x.c:63]: hmm a warning will be good I think

**Comment by plbossart** [comment]
@libinyang can we please complete the work on one platform before adding another PR for another. I can't keep up with all your changes and I still don't have a good idea of where you are heading. If you get silence as a reply to your questions, it' likely because no one understands them.

**Comment by libinyang** [comment]
@plbossart Yes, I'm working on UP2 of bxt_pcm512x platform. I will not submit PRs for other platform now.
I have sent an email to Takashi personally. He is OK that we set device_link in machine driver like this patch. (I will submit an patch for cAVS general machine driver for to community.) So now only this patch is enough for the HDMI power sequence issue. Together with the patches of https://github.com/thesofproject/linux/pull/725, the HDMI S3 issue can be fixed perfectly. And I have test on cAVS. Leafhill will fail in HDMI S3 test. And #725 patches and another patch of device_link patch in cAVS, HDMI S3 works.

> @libinyang can we please complete the work on one platform before adding another PR for another. I can't keep up with all your changes and I still don't have a good idea of where you are heading. If you get silence as a reply to your questions, it' likely because no one understands them.

**Comment by ranj063** [comment]
@libinyang this patch makes sense to me. But can send the upstream patch for cavs first before we merge this?

File: sound/soc/intel/boards/bxt_pcm512x.c [hunk 2]
```diff
@@ -53,6 +54,19 @@ static int broxton_hdmi_init(struct snd_soc_pcm_runtime *rtd)
...
+					    DL_FLAG_RPM_ACTIVE);
```
Comment (lyakh) [inline sound/soc/intel/boards/bxt_pcm512x.c:64]: Wondering, wouldn't adding a DL_FLAG_AUTOREMOVE_CONSUMER flag would eliminate the need for bxt_pcm512x_remove() below? Or maybe even also add DL_FLAG_AUTOREMOVE_SUPPLIER?

Comment (libinyang) [reply sound/soc/intel/boards/bxt_pcm512x.c:64]: @lyakh  I checked the DL_FLAG_AUTOREMOVE_xxx before. I found very little drivers are using these flags. What's more, looking into the code, what I understand is that we still need call some functions to release the link even we use the flags? Based on the result, I didn't add these flags for safe. And it seems Takashi is OK for using the flag DL_FLAG_RPM_ACTIVE.

**Comment by plbossart** [comment]
no longer needed with PR#851, closing

**Pull request closed without merging**